### PR TITLE
Gil - new launcher

### PR DESCRIPTION
**Pull request by GilStein1**
new launcher

### CODE REVIEW

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Bindings.java [hunk 1]
```diff
@@ -1,10 +1,18 @@
...
+
 import org.firstinspires.ftc.teamcode.gamepads.GamepadWrapper;
+import org.firstinspires.ftc.teamcode.subsystems.launcher.LauncherCommands;
 
 public class Bindings {
 
     private static GamepadWrapper mainGamepad;
     private static GamepadWrapper secondGamepad;
 
+    public static void razLauncherTestBindings(Gamepad gamepad1, Gamepad gamepad2) {
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Bindings.java]:
```suggestion
    public static void razLauncherTest(Gamepad gamepad1, Gamepad gamepad2) {
```

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Bindings.java [hunk 2]
```diff
@@ -1,10 +1,18 @@
...
+        mainGamepad = new GamepadWrapper(gamepad1);
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Bindings.java:20]: add line space

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Robot.java [hunk 3]
```diff
@@ -18,9 +20,11 @@ public static Robot getInstance() {
     }
 
     //add subsystems...
+    public Launcher launcher;
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Robot.java]: make private, and add get function in the bottom of the class

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmodes/RazLauncherTest.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package org.firstinspires.ftc.teamcode.opmodes;
+
+import org.firstinspires.ftc.teamcode.Bindings;
+import org.firstinspires.ftc.teamcode.Robot;
+
+public class RazLauncherTest extends DefaultRaz {
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmodes/RazLauncherTest.java]: missing annonatin

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/Launcher.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+                LauncherConstants.MIN_ANGLE,
+                LauncherConstants.MAX_ANGLE
+        );
+    }
+
+    public double getPosition() {
+        return servo.getPosition();
+    }
+
+    public void release() {
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/Launcher.java]: make protected, and change to launch plane

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/LauncherCommands.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
+package org.firstinspires.ftc.teamcode.subsystems.launcher;
+
+import com.arcrobotics.ftclib.command.Command;
+import com.arcrobotics.ftclib.command.InstantCommand;
+
+import org.firstinspires.ftc.teamcode.Robot;
+
+public class LauncherCommands {
+
+    public static Command launchFromLauncher() {
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/LauncherCommands.java]:
```suggestion
    public static Command launchPlane() {
```

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/LauncherConstants.java [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package org.firstinspires.ftc.teamcode.subsystems.launcher;
+
+public class LauncherConstants {
+    protected static final String SERVO_ID = "launcher_servo";
+    protected static final double RELEASED_POSITION = 0.95;
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/LauncherConstants.java]: make 1

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/Launcher.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        return getPosition() == LauncherConstants.RELEASED_POSITION;
+    }
+
+    public boolean isClosed() {
+        return getPosition() == LauncherConstants.CLOSED_POSITION;
+    }
+
+    public double getCurrentAngle() {
+        return getPosition() * (LauncherConstants.MAX_ANGLE - LauncherConstants.MIN_ANGLE) + LauncherConstants.MIN_ANGLE;
+    }
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/Launcher.java]: remove this

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/Launcher.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+    public Launcher(HardwareMap hardwareMap) {
+        this.servo = new SimpleServo(hardwareMap,
+                LauncherConstants.SERVO_ID,
+                LauncherConstants.MIN_ANGLE,
+                LauncherConstants.MAX_ANGLE
+        );
+    }
+
+    public double getPosition() {
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/Launcher.java]: make private

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmodes/RazLauncherTest.java
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmodes/RazLauncherTest.java:1]: put in package named testers

**Review by greenblitz4590 — COMMENTED** [review]
Nice Code!

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/Launcher.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
+package org.firstinspires.ftc.teamcode.subsystems.launcher;
+
+import com.arcrobotics.ftclib.command.SubsystemBase;
+import com.arcrobotics.ftclib.hardware.SimpleServo;
+import com.qualcomm.robotcore.hardware.HardwareMap;
+
+import org.firstinspires.ftc.robotcore.external.Telemetry;
+
+public class Launcher extends SubsystemBase {
+    private SimpleServo servo;
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/Launcher.java]: final

Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/Launcher.java]: make servo

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/Launcher.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+    public Launcher(HardwareMap hardwareMap) {
+        this.servo = new SimpleServo(hardwareMap,
+                LauncherConstants.SERVO_ID,
+                LauncherConstants.MIN_ANGLE,
+                LauncherConstants.MAX_ANGLE
+        );
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/Launcher.java]: servo constructur

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/Launcher.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+    private double getPosition() {
+        return servo.getPosition();
+    }
+
+    protected void launchPlane() {
+        servo.setPosition(LauncherConstants.RELEASED_POSITION);
+    }
+
+    public boolean isFullyReleased() {
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/Launcher.java]: is plane lauchned

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/Launcher.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+    protected void launchPlane() {
+        servo.setPosition(LauncherConstants.RELEASED_POSITION);
+    }
+
+    public boolean isPlaneLaunched() {
+        return getPosition() == LauncherConstants.RELEASED_POSITION;
+    }
+
+    public boolean isClosed() {
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/Launcher.java]: what is close

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/LauncherCommands.java [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+import com.arcrobotics.ftclib.command.Command;
+import com.arcrobotics.ftclib.command.InstantCommand;
+
+import org.firstinspires.ftc.teamcode.Robot;
+
+public class LauncherCommands {
+
+    public static Command launchPlane() {
+        return new InstantCommand(() -> Robot.getInstance().getLauncher().launchPlane(),Robot.getInstance().getLauncher());
+    }
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/LauncherCommands.java:12]: space before last line

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/LauncherConstants.java [hunk 15]
```diff
@@ -0,0 +1,7 @@
+package org.firstinspires.ftc.teamcode.subsystems.launcher;
+
+class LauncherConstants {
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/LauncherConstants.java:3]: line space after {

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/LauncherConstants.java [hunk 16]
```diff
@@ -0,0 +1,7 @@
...
+    protected static final String SERVO_ID = "launcher_servo";
+    protected static final double RELEASED_POSITION = 0;
+    protected static final double CLOSED_POSITION = 0.5;
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/LauncherConstants.java]: line space before last line

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/LauncherConstants.java [hunk 17]
```diff
@@ -0,0 +1,7 @@
+package org.firstinspires.ftc.teamcode.subsystems.launcher;
+
+class LauncherConstants {
+    protected static final String SERVO_ID = "launcher_servo";
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/LauncherConstants.java:5]: line space after this line

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/Launcher.java [hunk 18]
```diff
@@ -0,0 +1,38 @@
+package org.firstinspires.ftc.teamcode.subsystems.launcher;
+
+import com.arcrobotics.ftclib.command.SubsystemBase;
+import com.qualcomm.robotcore.hardware.HardwareMap;
+import com.qualcomm.robotcore.hardware.Servo;
+
+import org.firstinspires.ftc.robotcore.external.Telemetry;
+
+public class Launcher extends SubsystemBase {
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/Launcher.java:9]: line space after

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmodes/testers/RazLauncherTest.java [hunk 19]
```diff
@@ -0,0 +1,25 @@
+package org.firstinspires.ftc.teamcode.opmodes.testers;
+
+import org.firstinspires.ftc.teamcode.Bindings;
+import org.firstinspires.ftc.teamcode.Robot;
+import org.firstinspires.ftc.teamcode.opmodes.DefaultRaz;
+
+import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
+
+@TeleOp(name = "Raz Launch Test")
+public class RazLauncherTest extends DefaultRaz {
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmodes/testers/RazLauncherTest.java:10]: space

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmodes/testers/RazLauncherTest.java [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+    @Override
+    public void execute() {
+        Robot.getInstance().getLauncher().telemetry(telemetry);
+    }
+
+    @Override
+    public void configureBindings() {
+        Bindings.razLauncherTest(gamepad1, gamepad2);
+    }
+}
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmodes/testers/RazLauncherTest.java:27]: space

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/LauncherConstants.java [hunk 21]
```diff
@@ -0,0 +1,7 @@
...
+    protected static final double RELEASED_POSITION = 0;
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/LauncherConstants.java]: rename launch

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/LauncherConstants.java (hunk 16 above)
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/launcher/LauncherConstants.java]: rename held

**Review by YHerm — DISMISSED** [review]
Good Job!!!

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Robot.java [hunk 22]
```diff
@@ -20,22 +20,25 @@ public static Robot getInstance() {
...
-
     public void initSubsystems(HardwareMap hardwareMap) {
+        this.launcher = new Launcher(hardwareMap);
         this.arm = new Arm(hardwareMap);
         this.elevator = new Elevator(hardwareMap);
     }
 
+    private Arm arm;
+    private Elevator elevator;
+    private Launcher launcher;
```
Comment (greenblitz4590) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Robot.java]: why not above

**Review by YHerm — APPROVED** [review]

**Pull request merged**